### ISSUE TITLE
Faulty type checking for input to class decorator

### ISSUE THREAD

**Issue by DarkLight1337** [microsoft/pyright#3708]
**Describe the bug**
I am trying to type annotate a class decorator that only accepts classes that implement a certain protocol. However, the linter fails to show an error when passing an incompatible class unless I use a `TypeVar` to annotate the input class. Even then, the linter only correctly shows an error when I invoke the decorator through the function syntax; it still fails if I use the decorator syntax (`@`) to decorate the class.

**Minimal Reproducible Example**
 `dec0` and `dec1` are dummy class decorators. They are functionally the same, except that the type of the input class is annotated using a `TypeVar` in `dec0`, whereas it is annotated using `Type` in `dec1`. `MyProto` is the protocol that should be implemented by the input class.

`Good` and `Bad` classes are decorated to test the type checker. The `Good` classes implement `MyProto`, and as such, should pass the type check; conversely, the `Bad` classes do not implement `MyProto`, and as such, should fail the type check.

```python
from typing import Protocol, TypeVar, Type

class MyProto(Protocol):
    def foo(self, x: int) -> int: ...

TMyProtoMeta = TypeVar('TMyProtoMeta', bound=Type[MyProto])
def dec0(t: TMyProtoMeta) -> TMyProtoMeta:
    return t

def dec1(t: Type[MyProto]):
    return t

@dec0           # Expected: pass | Actual: pass
class Good0:
    def foo(self, x: int) -> int:
        return x

dec0(Good0)     # Expected: pass | Actual: pass

@dec1           # Expected: pass | Actual: pass
class Good1:
    def foo(self, x: int) -> int:
        return x

dec1(Good1)     # Expected: pass | Actual: pass

@dec0           # Expected: fail | Actual: pass
class Bad0:
    pass

dec0(Bad0)      # Expected: fail | Actual: fail

@dec1           # Expected: fail | Actual: pass
class Bad1:
    pass

dec1(Bad1)      # Expected: fail | Actual: pass

```

**Expected behavior**
Comments indicate the expected versus actual behaviour of whether type checking passes for that case.

As a sanity check, I also checked the code with `mypy`, which correctly shows an error for each case that is expected to fail.

**VS Code extension or command-line**
I am linting my code with Pylance v2022.7.30 on VSCode.

**Comment by erictraut** [comment]
Which version of pyright are you using? I'm not seeing the behavior you're reporting. I receive errors in the places where you are expecting errors.

**Comment by DarkLight1337** [comment]
I have type checking enabled in Pylance, but the errors are still not showing up. Please see the screenshot below.

![image](https://user-images.githubusercontent.com/44970335/179337656-5ecc75c6-95ba-4a02-8864-34954a4e5240.png)

I also tried setting type checking mode to "strict" but that did not work either.

**Comment by erictraut** [comment]
I'm able to repro the bug you're seeing in pylance 2022.7.30. It was fixed in pyright about a week and a half ago, and it no longer repros with the prerelease 2022.7.31, available as an "insiders" build.

**Comment by DarkLight1337** [comment]
I see, good to know that!
